cryptarchia/cryptarchia.py

```diff
@@ -131,40 +218,47 @@ def on_block(self, block: BlockHeader):
             return
 
         # check if the new block extends an existing chain
-        if self.try_extend_chains(block):
-            return
+        succeeded_in_extending_a_chain = self.try_extend_chains(block)
+        if not succeeded_in_extending_a_chain:
+            # we failed to extend one of the existing chains,
+            # therefore we might need to create a new fork
+            new_chain = self.try_create_fork(block)
+            if new_chain is not None:
+                self.forks.append(new_chain)
+            else:
+                # otherwise, we're missing the parent block
+                # in that case, just ignore the block
+                return
+
+        # We may need to switch forks, lets run the fork choice rule to check.
+        new_chain = Follower.fork_choice(self.local_chain, self.forks)
+
+        if new_chain == self.local_chain:
+            # we have not re-org'd therefore we can simply update our ledger state
+            # if this block extend our local chain
+            if self.local_chain.tip() == block:
+                self.ledger_state.apply(block)
+        else:
+            # we have re-org'd, therefore we must roll back out ledger state and
+            # re-apply blocks from the new chain
+            ledger_state = self.ledger_state_snapshot.copy()
```

Comment:
nit: it's not clear that `ledger_state_snapshot` corresponds to the genesis state

Also, we might keep a ledger state for each chain to avoid this kind of re-org stuff. A linear history without rollbacks is easier to understand, when changing fork you simply switch to the ledger state of that fork

Reply:
ledger_state_snapshot does not necessarily correspond with genesis state. It is the most recently finalized ledger_state.

every 3k/f slots, it's safe to take a snapshot because we are very confident that there won't be a fork that lasts longer than 3k/f slots long.

So really the full spec will have something like:

- ledger_state_finalized
- ledger_state_finalizing
- ledger_state_current

where when `ledger_state_finalizing` is older than 3k/f slots, it transitions to `ledger_state_finalized` and `ledger_state_current`is moved to `ledger_state_finalizing`

Reply:
yeah that makes sense, I'll try it with a ledger state per fork and see how that feels

Reply:
(discused offline, we'll drop the "snapshot" here and instead replay from genesis in the case of re-orgs)